broadinstitute/single_cell_portal_core
# Replacing Webpack(er) with Vite (SCP-4121)

SCP-4121.  This is ready for review, as it works well for me, Jon, and Jean (and there's a lot here, so it will be good to get started soon).  We should hold off on merging it until we can figure out what's going on with Emily's environment -- this doesn't build at all for her.  

The two concepts I think that are most important to learn:
1. In development mode, our Vite/React javascript will be loaded and execute *after* any inline scripts in our rails templates.  (the opposite will be true in production).  This required a number of changes, such as the new `renderComponent()` methods in both application.js files.  
2. Hot module reload operates on components.  Non-component exports from a component file (e.g. if a helper function or constant is exported too) will often force a page reload if anything in that file changes.  So I've started migrating some non-component exports to instead be set on the components themselves.  

A document outlining more detail is [here](https://docs.google.com/document/d/171bD5dgP1gIL5gMnQXorn-w6wCPvXKemxKqXlsMZDoc/edit#)

TO TEST:

1. stop any existing servers
2. checkout branch db-vite-build
4. run bundle install
5. run bundle exec vite install
6. run yarn install
7. run rails s in a terminal
8. run bin/vite dev in another terminal
9. go to https://localhost:3000/ and confirm homepage loads
10. click around to various parts of the application, ensuring functionality works.  Then make changes to some javascript files (e.g. StudySearchResult.jsx) and confirm live reload works.  There may be many files which still trigger a full page reload unnecessarily (see learning 2 above), but we can address these as we work on those files.
